### PR TITLE
refactor: change CompletionQueueImpl::cq return type

### PR DESCRIPTION
**Pull request by scotthart**
This change is in preparation for adding derivations of `CompletionQueueImpl` that do not contain a `grpc::CompletionQueue`.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451)

### CODE REVIEW

File: google/cloud/internal/completion_queue_impl.h [hunk 1]
```diff
@@ -71,8 +71,8 @@ class CompletionQueueImpl {
   virtual void StartOperation(std::shared_ptr<AsyncGrpcOperation> op,
                               absl::FunctionRef<void(void*)> start) = 0;
 
-  /// The underlying gRPC completion queue.
-  virtual grpc::CompletionQueue& cq() = 0;
+  /// The underlying gRPC completion queue, if it exists.
+  virtual grpc::CompletionQueue* cq() { return nullptr; }
```
Comment (devbww) [inline google/cloud/internal/completion_queue_impl.h:75]: Why not keep it pure?

File: google/cloud/internal/default_completion_queue_impl.cc [hunk 2]
```diff
@@ -217,7 +217,7 @@ DefaultCompletionQueueImpl::MakeDeadlineTimer(
     std::chrono::system_clock::time_point deadline) {
   auto p = AsyncTimerFuture::Create();
   auto op = std::move(p.first);
-  StartOperation(op, [&](void* tag) { op->Set(cq(), deadline, tag); });
+  StartOperation(op, [&](void* tag) { op->Set(*cq(), deadline, tag); });
```
Comment (coryan) [inline google/cloud/internal/default_completion_queue_impl.cc:220]: Maybe add a comment indicating this indirection is safe, as the value is returned from this class?  Should we mark the class `final` to make sure nobody overrides `cq()` to return `nullptr`?

File: generator/integration_tests/tests/golden_kitchen_sink_stub_test.cc [hunk 3]
```diff
@@ -478,7 +477,7 @@ TEST_F(GoldenKitchenSinkStubTest, AsyncStreamingWriteRead) {
 
   auto mock_cq = std::make_shared<MockCompletionQueueImpl>();
   grpc::CompletionQueue grpc_cq;
-  EXPECT_CALL(*mock_cq, cq).WillRepeatedly(ReturnRef(grpc_cq));
+  EXPECT_CALL(*mock_cq, cq).WillRepeatedly(Return(&grpc_cq));
```
Comment (coryan) [inline generator/integration_tests/tests/golden_kitchen_sink_stub_test.cc:480]: I suspect this test (and all the other tests in this file) can be changed to `Return(nullptr)`.

File: generator/integration_tests/tests/golden_thing_admin_stub_test.cc [hunk 4]
```diff
@@ -601,7 +600,7 @@ TEST_F(GoldenStubTest, AsyncCreateDatabase) {
         f(op.get());
         op->Notify(false);
       });
-  EXPECT_CALL(*mock, cq).WillOnce(ReturnRef(grpc_cq_));
+  EXPECT_CALL(*mock, cq).WillOnce(Return(&grpc_cq_));
```
Comment (coryan) [inline generator/integration_tests/tests/golden_thing_admin_stub_test.cc:603]: This file also seems amenable to `Return(nullptr)`.

File: google/cloud/bigtable/internal/bigtable_channel_refresh_test.cc [hunk 5]
```diff
@@ -75,13 +75,13 @@ TEST(BigtableChannelRefresh, Disabled) {
...
   using ns = std::chrono::nanoseconds;
-  using ::testing::ReturnRef;
+  using ::testing::Return;
 
   promise<StatusOr<std::chrono::system_clock::time_point>> p;
 
   auto mock_cq = std::make_shared<MockCompletionQueueImpl>();
   grpc::CompletionQueue grpc_cq;
-  EXPECT_CALL(*mock_cq, cq).WillRepeatedly(ReturnRef(grpc_cq));
+  EXPECT_CALL(*mock_cq, cq).WillRepeatedly(Return(&grpc_cq));
```
Comment (coryan) [inline google/cloud/bigtable/internal/bigtable_channel_refresh_test.cc:84]: Ditto, I do not think the `grpc_cq` value is used.

File: google/cloud/internal/async_read_write_stream_impl_test.cc [hunk 6]
```diff
@@ -32,7 +32,7 @@ namespace {
 using ::google::cloud::testing_util::IsOk;
 using ::google::cloud::testing_util::MockCompletionQueueImpl;
 using ::testing::_;
-using ::testing::ReturnRef;
+using ::testing::Return;
```
Comment (coryan) [inline google/cloud/internal/async_read_write_stream_impl_test.cc:35]: These remaining tests may be able to use `Return(nullptr)`, but I am less certain.

**Review by coryan — COMMENTED** [review]
The stuff around `final` and documenting why `*cq()` is safe seems important.  I would like to hear your thoughts on the suggestions for the tests, but I could be convinced either way.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451)* status: 0 of 17 files reviewed, 6 unresolved discussions (waiting on @coryan and @devbww)

___
*[`generator/integration_tests/tests/golden_kitchen_sink_stub_test.cc` line 480 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451#-NJNFdRHvbdONorJyBJS-r1-480:-NJO1KsT5TR94nXoXpX6:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9d151b21c65062ac204a92d6b0ce28462a00c44e/generator/integration_tests/tests/golden_kitchen_sink_stub_test.cc#L480)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I suspect this test (and all the other tests in this file) can be changed to `Return(nullptr)`.
</blockquote></details>

Done.
___
*[`generator/integration_tests/tests/golden_thing_admin_stub_test.cc` line 603 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451#-NJNFdRHvbdONorJyBJT-r1-603:-NJO1LqmB1EbaXcj1_1j:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9d151b21c65062ac204a92d6b0ce28462a00c44e/generator/integration_tests/tests/golden_thing_admin_stub_test.cc#L603)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

This file also seems amenable to `Return(nullptr)`.
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/internal/bigtable_channel_refresh_test.cc` line 84 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451#-NJNFdRHvbdONorJyBJU-r1-84:-NJO1MkxCwdc0XHQxbAp:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9d151b21c65062ac204a92d6b0ce28462a00c44e/google/cloud/bigtable/internal/bigtable_channel_refresh_test.cc#L84)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Ditto, I do not think the `grpc_cq` value is used.
</blockquote></details>

Done.
___
*[`google/cloud/internal/async_read_write_stream_impl_test.cc` line 35 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451#-NJNFdRHvbdONorJyBJZ-r1-35:-NJO1NbSDQdyvwnxLlFd:bj038o2) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9d151b21c65062ac204a92d6b0ce28462a00c44e/google/cloud/internal/async_read_write_stream_impl_test.cc#L35)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

These remaining tests may be able to use `Return(nullptr)`, but I am less certain.
</blockquote></details>

They all work with nullptr, updated accordingly.
___
*[`google/cloud/internal/completion_queue_impl.h` line 75 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451#-NJNFdRHvbdONorJyBJd-r1-75:-NJO1_SS1IB_unFIoSRE:bbm3xg4) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9d151b21c65062ac204a92d6b0ce28462a00c44e/google/cloud/internal/completion_queue_impl.h#L75)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Why not keep it pure?
</blockquote></details>

If it was pure, then derived types that don't have a grpc::CompletionQueue would need to override the method just to return nullptr? Seems unnecessary.
___
*[`google/cloud/internal/default_completion_queue_impl.cc` line 220 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451#-NJNFdRHvbdONorJyBJe-r1-220:-NJO1YWK1kFzlG2lkNmX:b-hma36g) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9d151b21c65062ac204a92d6b0ce28462a00c44e/google/cloud/internal/default_completion_queue_impl.cc#L220)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Maybe add a comment indicating this indirection is safe, as the value is returned from this class?  Should we mark the class `final` to make sure nobody overrides `cq()` to return `nullptr`?
</blockquote></details>

Added both.

**Review by devbww — COMMENTED** [review]
> Reviewed 10 of 17 files at r1, 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451)* status: all files reviewed, 6 unresolved discussions (waiting on @coryan and @scotthart)

___
*[`google/cloud/internal/completion_queue_impl.h` line 75 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451#-NJNFdRHvbdONorJyBJd-r1-75:-NJOvBY0Cml2QPl0e5ec:b-kj8p4x) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9d151b21c65062ac204a92d6b0ce28462a00c44e/google/cloud/internal/completion_queue_impl.h#L75)):*
> If it was pure, then derived types that don't have a grpc::CompletionQueue would need to override the method just to return nullptr

I would say, "exactly!"  That is, that sounds like a feature.  But I'm happy to yield to the consensus.

**Review by coryan — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451)* status: all files reviewed, 1 unresolved discussion (waiting on @devbww)

___
*[`google/cloud/internal/completion_queue_impl.h` line 75 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451#-NJNFdRHvbdONorJyBJd-r1-75:-NJQHZsDBd7hRwBmoTKB:b1i6x1q) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9d151b21c65062ac204a92d6b0ce28462a00c44e/google/cloud/internal/completion_queue_impl.h#L75)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

> If it was pure, then derived types that don't have a grpc::CompletionQueue would need to override the method just to return nullptr

I would say, "exactly!"  That is, that sounds like a feature.  But I'm happy to yield to the consensus.
</blockquote></details>

I did not think about it before, but keeping this class as a pure abstract class seems less confusing.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451)* status: all files reviewed, 1 unresolved discussion (waiting on @coryan and @devbww)

___
*[`google/cloud/internal/completion_queue_impl.h` line 75 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451#-NJNFdRHvbdONorJyBJd-r1-75:-NJRKk4MCjkBYC8RHXcT:b-7wdkz6) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9d151b21c65062ac204a92d6b0ce28462a00c44e/google/cloud/internal/completion_queue_impl.h#L75)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I did not think about it before, but keeping this class as a pure abstract class seems less confusing.
</blockquote></details>

So when it comes time to add `virtual rest_internal::RestCompletionQueue rest_cq() = 0;` that will also be what is desired?

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451)* status: all files reviewed, 1 unresolved discussion (waiting on @coryan and @devbww)

___
*[`google/cloud/internal/completion_queue_impl.h` line 75 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451#-NJNFdRHvbdONorJyBJd-r1-75:-NJRPZTa18MgYD8eOZpP:b-n7y0bd) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9d151b21c65062ac204a92d6b0ce28462a00c44e/google/cloud/internal/completion_queue_impl.h#L75)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Forgive my ignorance of the design, but what is the need for a `rest_cq()`?  Isn't `RestCompletionQueue` going to BE-A `CompletionQueueImpl`?
</blockquote></details>

The same way there is a grpc::CompletionQueue and a DefaultCompletionQueueImpl, there will be an analagous rest_internal::RestCompletionQueue and a RestComplectionQueueImpl.

**Review by devbww — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451)* status: all files reviewed, 1 unresolved discussion (waiting on @coryan)

___
*[`google/cloud/internal/completion_queue_impl.h` line 75 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451#-NJNFdRHvbdONorJyBJd-r1-75:-NJRSEXy8pkla0xpberz:b-xz6lx8) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9d151b21c65062ac204a92d6b0ce28462a00c44e/google/cloud/internal/completion_queue_impl.h#L75)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

The same way there is a grpc::CompletionQueue and a DefaultCompletionQueueImpl, there will be an analagous rest_internal::RestCompletionQueue and a RestComplectionQueueImpl.
</blockquote></details>

It seems to me that `cq()` is only a complication of having to interact with a foreign library (gRPC), and that we can avoid mirroring that hierarchy.  That is, I was thinking `RestCompletionQueueImpl` (aka `ThreadedCompletionQueueImpl`?) IS-A `CompletionQueueImpl`, returning null for `cq()`, and that's all.  But, you've thought through the details and I haven't, so I'll certainly defer.

**Review by coryan — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451)* status: all files reviewed, 1 unresolved discussion (waiting on @devbww)

___
*[`google/cloud/internal/completion_queue_impl.h` line 75 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451#-NJNFdRHvbdONorJyBJd-r1-75:-NJRV77yAdsp3DaFNZgg:b-9c3sgr) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9d151b21c65062ac204a92d6b0ce28462a00c44e/google/cloud/internal/completion_queue_impl.h#L75)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

It seems to me that `cq()` is only a complication of having to interact with a foreign library (gRPC), and that we can avoid mirroring that hierarchy.  That is, I was thinking `RestCompletionQueueImpl` (aka `ThreadedCompletionQueueImpl`?) IS-A `CompletionQueueImpl`, returning null for `cq()`, and that's all.  But, you've thought through the details and I haven't, so I'll certainly defer.
</blockquote></details>

If the time comes to add `rest_cq()`, then yes.  We should probably avoid it as @devbww describes, or using some kind of double dispatch / dynamic cast / wave-hands.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451)* status: all files reviewed, 1 unresolved discussion (waiting on @coryan and @devbww)

___
*[`google/cloud/internal/completion_queue_impl.h` line 75 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451#-NJNFdRHvbdONorJyBJd-r1-75:-NJRYYFn-beNGfcNCipI:bc905zg) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9d151b21c65062ac204a92d6b0ce28462a00c44e/google/cloud/internal/completion_queue_impl.h#L75)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

If the time comes to add `rest_cq()`, then yes.  We should probably avoid it as @devbww describes, or using some kind of double dispatch / dynamic cast / wave-hands.
</blockquote></details>

Considering this is all `internal`, we can iterate on these mechanisms as many times as care to. I'm merely trying to unblock myself.

**Comment by coryan** [comment]
*[`google/cloud/internal/completion_queue_impl.h` line 75 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451#-NJNFdRHvbdONorJyBJd-r1-75:-NJRZ4Nm-6fJgCp6qS_O:bp973rx) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9d151b21c65062ac204a92d6b0ce28462a00c44e/google/cloud/internal/completion_queue_impl.h#L75)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

Considering this is all `internal`, we can iterate on these mechanisms as many times as care to. I'm merely trying to unblock myself.
</blockquote></details>

Let's keep this function pure virtual, and we can discuss the other stuff later?

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451)* status: 16 of 17 files reviewed, 1 unresolved discussion (waiting on @coryan and @devbww)

___
*[`google/cloud/internal/completion_queue_impl.h` line 75 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451#-NJNFdRHvbdONorJyBJd-r1-75:-NJRcQox3n-GXGct1oJX:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9d151b21c65062ac204a92d6b0ce28462a00c44e/google/cloud/internal/completion_queue_impl.h#L75)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Let's keep this function pure virtual, and we can discuss the other stuff later?
</blockquote></details>

Done.

**Review by coryan — APPROVED** [review]
:lgtm:

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10451#-:-NJRixjTCjNP0OJtUrrf:bnfp4nl)* status: 16 of 17 files reviewed, 1 unresolved discussion (waiting on @devbww)

**Pull request merged**